src/pages/instances/InstanceConsole.tsx

```diff
@@ -130,13 +140,18 @@ const InstanceConsole: FC<Props> = ({ instance }) => {
           />
         </div>
       )}
-      {!isGraphic && (
+      {!isGraphic && canAccessInstanceConsole(instance) && (
         <InstanceTextConsole
           instance={instance}
           onFailure={onFailure}
           showNotRunningInfo={showNotRunningInfo}
         />
       )}
+      {!canAccessInstanceConsole(instance) && (
+        <Notification severity="caution" title="Restricted permissions">
+          You do not have permission to access the console for this instance.
+        </Notification>
+      )}
```

Comment:
I think this should be an early exit above the render function, keeping the rest of the logic above simpler.